### PR TITLE
feat: Gauntlet aera ark

### PR DESCRIPTION
**Pull request by halaprix**
# Pull Request: AeraArk Implementation with Async Deposit/Withdrawal Support

## Description
This PR implements a complete AeraArk contract that integrates with the Aera/Gauntlet provisioner system for async deposit and withdrawal operations. The implementation includes comprehensive test coverage with refund scenarios, proper authorization handling, and eliminates all magic values through named constants.

### Core Implementation
- **AeraArk Contract**: Complete implementation extending `ArkWithWithdrawalRequest` for Aera/Gauntlet Alpha vault integration
- **Async Operations**: Support for asynchronous deposits via `provisioner.requestDeposit()` and withdrawals via `provisioner.requestRedeem()`
- **Request Tracking**: State management for pending deposit/redeem requests using `ArkRequest` structs
- **Total Assets Calculation**: Accurate accounting including vault units, pending queues, and withdrawable assets

### Test Suite Enhancements
- **Helper Functions**: Added `_solveAsyncDepositRequest()` and `_solveAsyncRedeemRequest()` to simulate provisioner operations
- **Authorization Handling**: Proper use of `solveRequestsVault()` with provisioner owner authorization via Solmate Auth pattern
- **Price Manipulation**: `_increaseUnitPrice()` helper to handle rounding issues in redeem requests using accountant authorization
- **Comprehensive Refund Tests**: 
  - `test_DepositRefund_ExpiredRequest()` - Expired deposit refund via `refundRequest()`
  - `test_DepositRefund_DirectRefund()` - Direct deposit refund
  - `test_WithdrawalRefund_ExpiredRequest()` - Expired withdrawal refund
  - `test_TotalAssetsContinuity_MultipleScenarios()` - Asset continuity across multiple operations

### Constants and Code Quality
- **Named Constants**: Replaced all magic values with meaningful constants:
  - `REQUEST_DEADLINE = 24 hours` (public in contract, derived in tests)
  - `MAX_PRICE_AGE = 1 hours` 
  - `DEFAULT_SOLVER_TIP = 0`
  - `EXPIRY_BUFFER = 1 hours` (test-specific)
  - `PRICE_INCREASE_BPS = 10001` (0.01% for price manipulation)
- **Event Expectations**: Added `vm.expectEmit()` for proper event testing
- **Code Formatting**: Consistent formatting for long parameter lists and assertions

### Functionality
- **Complete Aera Integration**: Full support for Aera/Gauntlet Alpha vault operations
- **Async Request Handling**: Proper simulation of real-world async deposit/withdrawal flows
- **Robust Error Handling**: Comprehensive coverage of refund scenarios and edge cases
- **Accurate Asset Tracking**: Reliable total assets calculation across all states

### Code Quality
- **Maintainability**: Named constants make configuration changes simple and consistent
- **Readability**: Self-documenting code with meaningful variable names
- **Testability**: Comprehensive test coverage including edge cases and refund scenarios
- **Professional Standards**: No magic numbers, proper event testing, consistent formatting

### Correctness
- **Authorization**: Proper use of provisioner owner for `solveRequestsVault()` calls
- **Refund Mechanics**: Accurate implementation where USDC returns to ark (not commander) on deposit refunds
- **Rounding Handling**: Price manipulation workaround for provisioner's strict `AmountBoundExceeded` checks
- **Asset Continuity**: Verified total assets remain consistent across all operations

### Test Coverage
- ✅ **Basic Operations**: Constructor, board, total assets, vault units, harvest
- ✅ **Withdrawal Flow**: Request withdrawal, claim withdrawal, assets in queue
- ✅ **Async Operations**: Complete deposit and withdrawal flows with provisioner solving
- ✅ **Refund Scenarios**: Both expired and direct refunds for deposits and withdrawals
- ✅ **Edge Cases**: Asset continuity, authorization failures, rounding issues
- ✅ **Integration**: Real fork testing against Base mainnet Aera contracts

### Test Execution
```bash
# All AeraArk tests pass
forge test --match-contract AeraArkTestFork -v

# Specific test categories
forge test --match-test "test_Board" -v                    # Basic operations
forge test --match-test "test_CompleteAsyncWithdrawal" -v  # Full withdrawal flow  
forge test --match-test "Refund" -v                        # All refund scenarios
forge test --match-test "test_TotalAssetsContinuity" -v    # Asset continuity
```

### Key Test Insights
- **Authorization Discovery**: Found that `solveRequestsVault()` requires provisioner owner, not just any authorized address
- **Refund Behavior**: Confirmed USDC returns to ark contract on deposit refunds, not to original commander
- **Rounding Issues**: Identified and worked around `AmountBoundExceeded` errors in provisioner's unit-to-token conversion
- **Event Verification**: Proper testing of `DepositRefunded` and `RedeemRefunded` events

### Immediate Actions
- [ ] **Code Review**: Security review of async request handling and authorization patterns
- [ ] **Gas Optimization**: Review gas usage of helper functions and constants
- [ ] **Documentation**: Update contract documentation with async operation details

### Future Considerations
- [ ] **AeraArk Bug Fix**: Address potential issue where `requestWithdrawal()` uses original amount as `minTokensOut` instead of accounting for rounding
- [ ] **Enhanced Error Handling**: Consider more graceful handling of provisioner edge cases
- [ ] **Monitoring Integration**: Add events/metrics for tracking async request success rates
- [ ] **Multi-Asset Support**: Extend pattern to other Aera vault types if needed

### Integration Tasks
- [ ] **Deployment Scripts**: Create deployment scripts with proper constructor parameters
- [ ] **Configuration Management**: Set up environment-specific constants
- [ ] **Frontend Integration**: Update UI to handle async deposit/withdrawal states
- [ ] **Monitoring Setup**: Dashboard for tracking pending requests and refund rates

### Technical Discoveries
- **Provisioner Architecture**: Two distinct refund mechanisms (`refundDeposit` for sync, `refundRequest` for async)
- **Authorization Pattern**: Aera uses Solmate Auth with separate `owner` and `authority` addresses
- **Price Manipulation**: Accountant address (`0x854e4f646a73CC34e89aD9E4826f57DCB265c77b`) can update unit prices for testing
- **Request Types**: Auto-price vs fixed-price requests have different solving mechanisms

### Design Decisions
- **Constants Strategy**: Made contract constants `public` for test derivation, avoiding duplication
- **Helper Functions**: Extracted complex provisioner interactions into reusable test helpers
- **Error Simulation**: Used price manipulation rather than modifying core contract logic for rounding issues
- **Event Testing**: Comprehensive event verification for all refund operations

### Potential Improvements
- **Request ID Tracking**: Current `withdrawalRequestId()` returns 0 - could implement proper tracking
- **Batch Operations**: Could extend to handle multiple concurrent requests
- **Deadline Management**: Could make request deadlines configurable per operation
- **Slippage Protection**: Could add configurable slippage bounds for price-sensitive operations

Please review and provide any feedback or suggestions for improvement.

### CODE REVIEW

File: packages/core-contracts/test/arks/AeraArk.t.sol [hunk 1]
```diff
@@ -0,0 +1,1254 @@
...
+        // Claim withdrawal (no-op for Aera since it's automatic)
+        vm.prank(keeper);
+        ark.claimWithdrawal(); // Should not revert
+
+        // Check that no claim is required
+        bool claimRequired = ark.isWithdrawalClaimRequired();
+        assertFalse(claimRequired, "Aera should not require manual claim");
+    }
+
+    function test_RequestWithdrawalXX() public {
```
Comment (zerotucks) [inline packages/core-contracts/test/arks/AeraArk.t.sol:593]: Think this is a typo

File: packages/core-contracts/src/contracts/arks/AeraArk.sol [hunk 2]
```diff
@@ -0,0 +1,384 @@
...
+        uint256 shareAtTheTimeOfDeposit = priceCalculator.convertTokenToUnits(
+            address(vault),
+            config.asset,
+            amount
+        );
+        config.asset.forceApprove(address(provisioner), amount);
+        provisioner.requestDeposit(
+            config.asset,
+            amount,
+            shareAtTheTimeOfDeposit, // minUnitsOut - calculated by solver
```
[the remark this answers is not in the document]
Comment (halaprix) [reply packages/core-contracts/src/contracts/arks/AeraArk.sol:157]: it's rather the other way around - if price goes down - then we'd get more units so that's fine. But maybe we coudl add the protection in the other direction - roight now it's forcing the solver to deposit before the next price update

File: packages/core-contracts/src/contracts/arks/AeraArk.sol [hunk 3]
```diff
@@ -0,0 +1,390 @@
...
+            );
+        }
+
+        if (sharesToRedeem == 0) return;
+
+        vault.forceApprove(address(provisioner), sharesToRedeem);
+        // Create async redeem request
+        provisioner.requestRedeem(
+            config.asset,
+            sharesToRedeem,
```
Comment (fico23) [inline packages/core-contracts/src/contracts/arks/AeraArk.sol:275]: since share price usually goes up(and you get more than you specified here), this would be fine for most cases. but it may still occasionally go down, so i'd still suggest some margin for error, to avoid having your requests stuck waiting for deadline to pass

**Review by fico23 — COMMENTED** [review]
aera dev here, wanted to directly comment on the PR to avoid message forwarder :)

generally looks good, but i've left some comments.

File: packages/core-contracts/src/contracts/arks/AeraArk.sol [hunk 4]
```diff
@@ -0,0 +1,393 @@
...
+        }
+
+        provisioner = IProvisioner(_provisioner);
+        vault = IERC20(provisioner.MULTI_DEPOSITOR_VAULT());
+        priceCalculator = IPriceAndFeeCalculator(
+            provisioner.PRICE_FEE_CALCULATOR()
+        );
+
+        // Approve the provisioner to spend the Ark's tokens
+        config.asset.forceApprove(_provisioner, Constants.MAX_UINT256);
```
[the remark this answers is not in the document]
Comment (halaprix) [reply packages/core-contracts/src/contracts/arks/AeraArk.sol:88]: Not needed at all, good catch! :sweat:  Now, let's test on prod. ;)
